assignment-section-05/classification_model/processing/data_manager.py

```diff
@@ -0,0 +1,103 @@
+import typing as t
+from pathlib import Path
+
+import joblib
+import re
+import numpy as np
+import pandas as pd
+from sklearn.pipeline import Pipeline
+
+from classification_model import __version__ as _version
+from classification_model.config.core import DATASET_DIR, TRAINED_MODEL_DIR, config
+
+
+# TODO: type hints
+def get_first_cabin(row):
+    try:
+        return row.split()[0]
+    except AttributeError:
+        return np.nan
+
+
+def get_title(passenger: str) -> str:
+    """Extracts the title (Mr, Ms, etc) from the name variable."""
+    line = passenger
+    if re.search("Mrs", line):
+        return "Mrs"
+    elif re.search("Mr", line):
+        return "Mr"
+    elif re.search("Miss", line):
+        return "Miss"
+    elif re.search("Master", line):
+        return "Master"
+    else:
+        return "Other"
+
+
+def pre_pipeline_preparation(*, dataframe: pd.DataFrame) -> pd.DataFrame:
+    # replace question marks with NaN values
+    data = dataframe.replace("?", np.nan)
+
+    # retain only the first cabin if more than
+    # 1 are available per passenger
+    data["cabin"] = data["cabin"].apply(get_first_cabin)
+
+    data["title"] = data["name"].apply(get_title)
+
+    # cast numerical variables as floats
+    data["fare"] = data["fare"].astype("float")
+    data["age"] = data["age"].astype("float")
+
+    # drop unnecessary variables
+    data.drop(labels=config.model_config.unused_fields, axis=1, inplace=True)
+
+    return data
+
+
+def _load_raw_dataset(*, file_name: str) -> pd.DataFrame:
+    dataframe = pd.read_csv(Path(f"{DATASET_DIR}/{file_name}"))
```

Comment:
I did not see the url/link to the online data anywhere. Is this because you ask students somewhere to download the data and store it as raw.csv?

Or maybe I missed the online link?

If you want to start by loading raw.csv, you need to make crystal clear that they need to download the data from the url and save it as raw.csv at the very beginning of the assignment. Otherwise, the students will try to pick it up from the URL as I do in the notebook.